client/fetch_tagged_results_accumulator.go

```diff
@@ -0,0 +1,366 @@
+	pools fetchTaggedPools,
+	elems fetchTaggedIDResults,
+) encoding.SeriesIterator {
+	numElems := len(elems)
+	iters := pools.IteratorArray().Get(numElems)[:numElems]
+	for idx, elem := range elems {
+		slicesIter := pools.ReaderSliceOfSlicesIterator().Get()
+		slicesIter.Reset(elem.Segments)
+		multiIter := pools.MultiReaderIterator().Get()
+		multiIter.ResetSliceOfSlices(slicesIter)
+		iters[idx] = multiIter
+	}
+
+	// pick the first element as they all have identical ids/tags
+	// NB: safe to assume this element exists as it's only called within
+	// a forEachID lambda, which provides the guarantee that len(elems) != 0
+	elem := elems[0]
+
+	encodedTags := pools.CheckedBytesWrapper().Get(elem.EncodedTags)
+	decoder := pools.TagDecoder().Get()
+	decoder.Reset(encodedTags)
+
+	tsID := pools.CheckedBytesWrapper().Get(elem.ID)
+	nsID := pools.CheckedBytesWrapper().Get(elem.NameSpace)
+	seriesIter := pools.SeriesIterator().Get()
+	seriesIter.Reset(pools.ID().BinaryID(tsID), pools.ID().BinaryID(nsID),
+		decoder, accum.startTime, accum.endTime, iters)
+
+	return seriesIter
+}
+
+func (accum *fetchTaggedResultAccumulator) AsEncodingSeriesIterators(
+	limit int, pools fetchTaggedPools,
+) (encoding.SeriesIterators, bool, error) {
+	results := fetchTaggedIDResultsSortedByID(accum.responses)
+	sort.Sort(results)
+	accum.responses = fetchTaggedIDResults(results)
+
+	numElements := 0
+	accum.responses.forEachID(func(_ fetchTaggedIDResults) bool {
+		numElements++
+		return numElements < limit
+	})
+
+	result := pools.MutableSeriesIterators().Get(numElements)
+	result.Reset(numElements)
+	count := 0
+	accum.responses.forEachID(func(elems fetchTaggedIDResults) bool {
+		seriesIter := accum.sliceResponsesAsSeriesIter(pools, elems)
+		result.SetAt(count, seriesIter)
+		count++
+		return count < limit
+	})
+
+	exhaustive := accum.exhaustive && count < limit
+	return result, exhaustive, nil
+}
+
+func (accum *fetchTaggedResultAccumulator) AsIndexQueryResults(
+	limit int,
+	pools fetchTaggedPools,
+) (index.QueryResults, error) {
+	var (
+		iter  = newFetchTaggedResultsIndexIterator(pools)
+		count = 0
+	)
+	results := fetchTaggedIDResultsSortedByID(accum.responses)
+	sort.Sort(results)
+	accum.responses = fetchTaggedIDResults(results)
+	accum.responses.forEachID(func(elems fetchTaggedIDResults) bool {
+		iter.backing.ids = append(iter.backing.ids, elems[0].ID)
+		iter.backing.nses = append(iter.backing.nses, elems[0].NameSpace)
+		iter.backing.tags = append(iter.backing.tags, elems[0].EncodedTags)
+		count++
+		return count < limit
+	})
+
+	return index.QueryResults{
+		Exhaustive: accum.exhaustive && count < limit,
+		Iterator:   iter,
+	}, nil
+}
+
+type fetchTaggedShardConsistencyResults []fetchTaggedShardConsistencyResult
+
+func (res fetchTaggedShardConsistencyResults) initialize(length int) fetchTaggedShardConsistencyResults {
+	if cap(res) < length {
+		res = append(res, make(fetchTaggedShardConsistencyResults, length-cap(res))...)
```

Comment:
Hm, this is kind of gross making a temporary slice only to throw it away likely (if you have to grow it).  You're not looking to keep the original slice contents around and it only ever gets called once at initialization yeah?  Perhaps just reallocate by hand?

i.e.
```go
if cap(res) < length {
  res = make(fetchTaggedShardConsistencyResults, length)
}
```

Reply:
Yeah that's completely fair. I'm doing an alloc anyway, much easier to read if do the full alloc I need. Yep, it'll only happen if max shards change ever.